zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/authorization/RoleMultiPartitionTest.java

```diff
@@ -0,0 +1,133 @@
+        .containsSubsequence(
+            tuple(CommandDistributionIntent.DISTRIBUTING, RecordType.EVENT, 2),
+            tuple(CommandDistributionIntent.ACKNOWLEDGE, RecordType.COMMAND, 2),
+            tuple(CommandDistributionIntent.ACKNOWLEDGED, RecordType.EVENT, 2))
+        .containsSubsequence(
+            tuple(CommandDistributionIntent.DISTRIBUTING, RecordType.EVENT, 3),
+            tuple(CommandDistributionIntent.ACKNOWLEDGE, RecordType.COMMAND, 3),
+            tuple(CommandDistributionIntent.ACKNOWLEDGED, RecordType.EVENT, 3))
+        .endsWith(tuple(CommandDistributionIntent.FINISHED, RecordType.EVENT, 1));
+    for (int partitionId = 2; partitionId < PARTITION_COUNT; partitionId++) {
+      assertThat(
+              RecordingExporter.roleRecords()
+                  .withPartitionId(partitionId)
+                  .limit(record -> record.getIntent().equals(RoleIntent.CREATED))
+                  .collect(Collectors.toList()))
+          .extracting(Record::getIntent)
+          .containsExactly(RoleIntent.CREATE, RoleIntent.CREATED);
+    }
+  }
+
+  @Test
+  public void shouldDistributeInIdentityQueue() {
+    // when
+    final var name = UUID.randomUUID().toString();
+    ENGINE.role().newRole(name).create();
+
+    // then
+    assertThat(
+            RecordingExporter.commandDistributionRecords()
+                .limitByCount(r -> r.getIntent().equals(CommandDistributionIntent.FINISHED), 1)
+                .withIntent(CommandDistributionIntent.ENQUEUED))
+        .extracting(r -> r.getValue().getQueueId())
+        .containsOnly(DistributionQueue.IDENTITY.getQueueId());
+  }
+
+  @Test
+  public void distributionShouldNotOvertakeOtherCommandsInSameQueue() {
+    // given the user creation distribution is intercepted
+    for (int partitionId = 2; partitionId <= PARTITION_COUNT; partitionId++) {
+      interceptRoleCreateForPartition(partitionId);
+    }
+
+    // when
+    final var name = UUID.randomUUID().toString();
+    ENGINE.role().newRole(name).create();
+
+    // Increase time to trigger a redistribution
+    ENGINE.increaseTime(Duration.ofMinutes(1));
+
+    // then
+    assertThat(
+            RecordingExporter.commandDistributionRecords(CommandDistributionIntent.FINISHED)
+                .limit(1))
+        .extracting(r -> r.getValue().getValueType(), r -> r.getValue().getIntent())
+        .containsExactly(Assertions.tuple(ValueType.ROLE, RoleIntent.CREATE));
+  }
+
+  private void interceptRoleCreateForPartition(final int partitionId) {
+    final var hasInterceptedPartition = new AtomicBoolean(false);
+    ENGINE.interceptInterPartitionCommands(
+        (receiverPartitionId, valueType, intent, recordKey, command) -> {
+          if (hasInterceptedPartition.get()) {
+            return true;
+          }
+          hasInterceptedPartition.set(true);
+          return !(receiverPartitionId == partitionId && intent == RoleIntent.CREATE);
+        });
+  }
```

Comment:
❌ The test setup is wrong this way. You don't test distribution ordering this way, you only test redistribution of the command. We should change this so we first put something on the identity queue (eg, create a user). Intercept the acknowledgement of this. Then when we create the role we must verify it doesn't get distributed immediately, but it is processed after the user create is retried:

```java
  @Test
  public void distributionShouldNotOvertakeOtherCommandsInSameQueue() {
    // given the user creation distribution is intercepted
    for (int partitionId = 2; partitionId <= PARTITION_COUNT; partitionId++) {
      interceptUserCreateForPartition(partitionId);
    }
    ENGINE
        .user()
        .newUser(UUID.randomUUID().toString())
        .withName("Foo Bar")
        .withPassword("baz")
        .withEmail("foobar@baz.com")
        .create();

    // when
    final var name = UUID.randomUUID().toString();
    ENGINE.role().newRole(name).create();

    // Increase time to trigger a redistribution
    ENGINE.increaseTime(Duration.ofMinutes(1));

    // then
    assertThat(
            RecordingExporter.commandDistributionRecords(CommandDistributionIntent.FINISHED)
                .limit(2))
        .extracting(r -> r.getValue().getValueType(), r -> r.getValue().getIntent())
        .containsExactly(
            tuple(ValueType.USER, UserIntent.CREATE), tuple(ValueType.ROLE, RoleIntent.CREATE));
  }

  private void interceptUserCreateForPartition(final int partitionId) {
    final var hasInterceptedPartition = new AtomicBoolean(false);
    ENGINE.interceptInterPartitionCommands(
        (receiverPartitionId, valueType, intent, recordKey, command) -> {
          if (hasInterceptedPartition.get()) {
            return true;
          }
          hasInterceptedPartition.set(true);
          return !(receiverPartitionId == partitionId && intent == RoleIntent.CREATE);
        });
  }
```

Reply:
You are right, sorry for missing this 🙏